### PR TITLE
cmake: implement build infrastructure for supporting SCA tools.

### PR DESCRIPTION
**Pull request by tejlmand**
Static code analyser (SCA) tools are important in software development.

CMake offers built-in support for some tools, such as cppcheck and
clang-tidy.

Other tools, such as sparse, are not directly supported.

This commit provides a uniform way for users to specify a supported
SCA using `ZEPHYR_SCA_VARIANT=<tool>` which is consistent with how
toolchains are specified.
ZEPHYR_SCA_VARIANT can be set using `-D` or in environment.

Support for an SCA tool is done in `cmake/sca/<tool>/sca.cmake`.
SCA_ROOT can be used to specify additional search paths when looking up
implementation for a tool. SCA_ROOT can also be specified in
`zephyr/module.yml` as setting. This makes it possible to provide SCA
tool implementation as part of a Zephyr module.

As a show-case, sparse support has been reworked into the new framework and old way of enabling sparse has been deprecated.

This PR is out-come of the Zephyr Architecture meeting, 2022-11-30.

--------

Pending work:
- [x] Create documentation and update existing Zephyr module docs to describe `sca_root` in module.yml.
- [x] Update Sparse doc: https://docs.zephyrproject.org/latest/develop/test/sparse.html
- [x] Decide if `SCA` is the proper name, or we want to use a different name for this functionality

### CODE REVIEW

**Comment by tejlmand** [comment]
This PR uses sca as name for variable and implementation, but i'm open to discuss if a better name can be found.

Also it introduced `ZEPHYR_SCA_VARIANT`, either as CMake variable or env setting, but another option would be to allow specifying the sca tool together with the toolchain variant, for example:
`ZEPHYR_TOOLCHAIN_VARIANT="zephyr;sparse"`.

I prefer two distinct variables, but I do see pros in being able to specify them together similar to CMake's extra generator, like: `Eclipse CDT4 - Ninja`.
Comments are welcome.

**Comment by pdgendt** [comment]
> This PR is out-come of the Zephyr Architecture meeting, 2022-11-30.

More info @ https://github.com/zephyrproject-rtos/zephyr/pull/51212#issuecomment-1330935540

cc @romkell

**Comment by tejlmand** [comment]
> It seems to be doing something, but fails on including toolchains.

Setting it in the CMake cache works, so try this implementation instead:
```
find_program(CPPCHECK_TOOL cppcheck)
set(CMAKE_C_CPPCHECK
    ${CPPCHECK_TOOL}
    --enable=warning
    --inconclusive
    --force
    --inline-suppr
    CACHE INTERNAL ""
)
```

gives this:
```
...
[8/163] Building C object zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj
Checking /projects/github/ncs/zephyr/arch/arm/core/offsets/offsets.c ...
[10/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/dec.c.obj
Checking /projects/github/ncs/zephyr/lib/os/dec.c ...
Checking /projects/github/ncs/zephyr/lib/os/dec.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;_ASMLANGUAGE...
[11/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/hex.c.obj
Checking /projects/github/ncs/zephyr/lib/os/hex.c ...
Checking /projects/github/ncs/zephyr/lib/os/hex.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;_ASMLANGUAGE...
[12/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/sem.c.obj
Checking /projects/github/ncs/zephyr/lib/os/sem.c ...
[13/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/printk.c.obj
Checking /projects/github/ncs/zephyr/lib/os/printk.c ...
[14/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/fdtable.c.obj
Checking /projects/github/ncs/zephyr/lib/os/fdtable.c ...
[15/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/timeutil.c.obj
Checking /projects/github/ncs/zephyr/lib/os/timeutil.c ...
Checking /projects/github/ncs/zephyr/lib/os/timeutil.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1...
[16/163] Building C object CMakeFiles/app.dir/src/main.c.obj
Checking /projects/github/ncs/zephyr/samples/hello_world/src/main.c ...
Checking /projects/github/ncs/zephyr/samples/hello_world/src/main.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;CONFIG_ARC;_ASMLANGUAGE...
Checking /projects/github/ncs/zephyr/samples/hello_world/src/main.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;CONFIG_ARCH_POSIX;_ASMLANGUAGE...
Checking /projects/github/ncs/zephyr/samples/hello_world/src/main.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;CONFIG_ARM64;_ASMLANGUAGE...
...
Checking /projects/github/ncs/zephyr/samples/hello_world/src/main.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;_ASMLANGUAGE...
Checking /projects/github/ncs/zephyr/samples/hello_world/src/main.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;_ASMLANGUAGE;__ASSEMBLER__...
[18/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/cbprintf_packaged.c.obj
Checking /projects/github/ncs/zephyr/lib/os/cbprintf_packaged.c ...
[19/163] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/rb.c.obj
Checking /projects/github/ncs/zephyr/lib/os/rb.c ...
Checking /projects/github/ncs/zephyr/lib/os/rb.c: KERNEL=1;NRF52840_XXAA=1;__PROGRAM_START=1;__ZEPHYR__=1;CONFIG_ARC;_ASMLANGUAGE...
...
```

**Comment by nashif** [comment]
we will get this working, my point was showing how easy it is to add tools, especially those already supported by cmake.

**Review by mbolivar-nordic — DISMISSED** [review]

File: cmake/modules/FindDeprecated.cmake [hunk 1]
```diff
@@ -92,5 +92,16 @@ if("SOURCES" IN_LIST Deprecated_FIND_COMPONENTS)
...
 endif()
 
+if("SPARSE" IN_LIST Deprecated_FIND_COMPONENTS)
+  # This code was deprecated after Zephyr v3.2.0
+  if(SPARSE)
+    message(DEPRECATION
+        "Setting SPARSE=${SPARSE} is deprecated. "
+        "Please set ZEPHYR_SCA_VARIANT to 'sparse'"
+    )
+    set_ifndef(ZEPHYR_SCA_VARIANT sparse)
```
Comment (marc-hb) [inline cmake/modules/FindDeprecated.cmake:102]: Thanks for that, "zero days" are not fun.

**Review by marc-hb — COMMENTED** [review]
I lack time and expertise for a detailed review of the CMake implementation details but the user interface looks good and the amount of code is surprisingly small for all this does, impressive! Thanks.

**Comment by nashif** [comment]
with the above cppcheck configuration and enabling misra addon (--addon=misra) I am now getting actual results...

```
Checking /home/nashif/zephyrproject/zephyr/kernel/version.c ...
Checking /home/nashif/zephyrproject/zephyr/kernel/version.c: KERNEL=1;_POSIX_C_SOURCE=200809;_XOPEN_SOURCE=600;_XOPEN_SOURCE_EXTENDED=1;__ZEPHYR_SUPERVISOR__=1;__ZEPHYR__=1...
/home/nashif/zephyrproject/zephyr/kernel/version.c:18:10: style: A compatible declaration shall be visible when an object or function with external linkage is defined  [misra-c2012-8.4]
uint32_t sys_kernel_version_get(void)
         ^
/home/nashif/zephyrproject/zephyr/include/zephyr/types.h:22:9: style: The union keyword should not be used  [misra-c2012-19.2]
typedef union {
        ^
/home/nashif/zephyrproject/zephyr/build/zephyr/include/generated/version.h:2:0: style: #define and #undef shall not be used on a reserved identifier or reserved macro name  [misra-c2012-21.1]
#define _KERNEL_VERSION_H_
^

```

**Review by nashif — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

**Comment by nashif** [comment]
need @lyakh  to confirm if sparse is working as before.

**Comment by tejlmand** [comment]
`DNM` to give time for @lyakh to respond.

**Review by nordicjm — DISMISSED** [review]

**Comment by romkell** [comment]
I will reflect what I read and understood from the code. The text may server as basis for a documentation.

=== start doc text ===

Static Code Analyzer work differently and depending on the tasks they perform, require more or less build information.
The following, possibly incomplete list of, constellations are known:
1. Stand-alone SCA's execution: during a build an information file (e.g. compile_commands.json) is generated. Those information then are used to launch the SCA. Build and SCA runs execute separately. An integration into the build system is not required and done. Typically the linking step cannot be considered in that constellation.
2. SCA execution during build
  2.1. The compiler / linker call is replaced by a call to the SCA in the build system. The SCA then calls the compiler after executing itself. The path to the compiler executable and all compiler command line parameter need to be known to the SCA. The build system needs to be modified to replace the compiler by the SCA call.
  2.2. The compiler / linker call is intercepted on an OS level (similar to a virus scanner). The SCA executes itself with all the command line parameters that were passed to the compiler call and then returns control back to the compiler executable. The SCA requires to know the path to compiler executable to be intercepted.

In the Zepyhr OS build system the SCA integration has been solved as follows:

- The variable `ZEPHYR_SCA_VARIANT` selects the SCA tool to be used during the build process
  - Pass `ZEPHYR_SCA_VARIANT` when starting a build either directly to cmake or otherwise west using `-DZEPHYR_SCA_VARIANT=\<sca-variant-name\>`
  - \<sca-variant-name\> has to correspond with one of the folder names in /cmake/sca/ e.g. 
   /cmake/sca/**sparse** or 
   /cmake/sca/**cppcheck**
- The variable `SCA_ROOT` sets the root paths for your out-of-tree configuration of a SCA tool.
  - `SCA_ROOT` may be a list of several locations/paths separated by either semi-colons or spaces for your out of tree sca tool configurations
  - Since a SCA configuration is executed from for all paths in `SCA_ROOT`, the last entry in `SCA_ROOT` has the highest priority
  - If there are several configurations for the same `ZEPHYR_SCA_VARIANT` name in different locations all preceding will get overwritten by the last found, also the in-tree configuration
  - Empty (non existing) locations of configurations for `ZEPHYR_SCA_VARIANT` will be ignored.
  - A custom out of tree sca configurations need to comply with the following structure:
        \<root-location(s)\>/cmake/sca/\<sca-variant-name\>/sca.make
	optional: \<root-location(s)\>/cmake/sca/\<sca-variant-name\>/\<sca-variant-name\>_\<lang\>_lancher.tmplt
	- sca.make: You have the freedom to implement what ever is required to get your SCA configured and integrated. sca.make will only be called once in a build run and **not** repeatedly when compiling the sources. Therefore sca.make is the right place to do once per build configurations of the SCA such as:
	  - writing SCA configuration files formatted according to your SCA's preference (consider cmake's https://cmake.org/cmake/help/latest/command/configure_file.html), 
	  - preparing and installing a launcher script which will call the SCA (see below)
	  - \<sca-variant-name\>\_\<lang\>\_launcher.tmplt: SCAs which are called as a replacement of the compiler and call the actual compiler after executing themself. Best use cmake's `CMAKE_<LANG>_COMPILER_LAUNCHER` feature (see https://cmake.org/cmake/help/latest/prop_tgt/LANG_COMPILER_LAUNCHER.html). For that purpose \<sca-variant-name\>\_\<lang\>\_launcher.tmplt can be implemented, which will be called by cmake as a replacement for the compiler itself. Preferably the \<sca-variant-name\>\_\<lang\>\_launcher.tmplt is a cmake script, but could be any other script / executable that runs in your environment (OS, installed tools)
	    - Once per build configurations are best not be done in the launcher script\(s\), since the script\(s\) will be called repeatedly for each source file.
	  
Please take the "sparce" (tbd: add link) integration as a simple example for case 2.1 and a good starting point for your own SCA integration.

=== end doc text ===
  
If I mirrored the implementation incorrectly somewhere, I would be glad to know.
Feel free to correct the text if you think it can be used as documentation basis but is not understandable enough.

We have not integrated our companies SCA (Axivion Suite) yet, but I am quite positive that it will work for us and for most.

I would also rename \<sca-variant-name\>.template to \<sca-variant-name\>_\<lang\>_launcher.tmplt or similar.
1. There may be several launcher: for C and C++ and the linker
2. If it is a launcher I would name it that way somehow. If there is a common understandable abbreviation for launcher also that could be used to shorten the file name.

**Comment by marc-hb** [comment]
BTW I just learned the hard way that `ccache` breaks `-DSPARSE=y` on Arch Linux. Ubuntu 22 is fine. No idea why. More specifically `ccache cgcc -o zephyr/zephyr_pre0.elf` fails with a gazillion errors like `first defined here` and other `zephyr/include/zephyr/arch/xtensa/atomic_xtensa.h:35: undefined reference to `xtensa_switch'`. Not sure why yet.

100% reproducible with sof/west.yml commit 4a9a99aeb24a

Obvious workaround: uninstall ccache.

cc: @andyross

**Comment by tejlmand** [comment]
> BTW I just learned the hard way that ccache breaks -DSPARSE=y on Arch Linux.

@marc-hb are we talking this PR or current Sparse implementation on zephyr/main ?
Or both cases ?

> More specifically `ccache cgcc -o zephyr/zephyr_pre0.elf` fails with a gazillion errors like first defined here and other `zephyr/include/zephyr/arch/xtensa/atomic_xtensa.h:35: undefined reference to 'xtensa_switch'`. Not sure why yet.

If running this by hand, then remember to have the right gcc set as `REAL_CC` in environment.
I assume you do, but just mention it to be on the safe side.

> Obvious workaround: uninstall ccache.

Or build with:
```
$ cmake ... -DUSE_CCACHE=0
```
or with `west`
```
$ west build ... -- -DUSE_CCACHE=0
```

**Comment by marc-hb** [comment]
> @marc-hb are we talking this PR or current Sparse implementation on zephyr/main ?

Fails to build only with the older SPARSE=y!! So this PR fixes a very strange bug in a very specific Arch Linux specific combination that I haven't even filed yet. I will probably save myself some time then :-) I was afraid to be off-topic, looks like I'm not after all.

> If running this by hand,

No I just copied the command that was failing, I didn't run it directly

> $ cmake ... -DUSE_CCACHE=0

This avoids the link failure too.

**Comment by tejlmand** [comment]
>  Not with this PR.

Maybe I should update one of the commit messages with:
```
Fixes: #marc-hb's problem
```
😆

**Comment by tejlmand** [comment]
> Fails to link only with the older SPARSE=y!

Thanks for that info, because then without examine the code itself then I have a big suspicion it's directly related to this comment:
https://github.com/zephyrproject-rtos/zephyr/pull/43776#discussion_r909458933

Because there are places in code with the sparse method introduced in #43776 there are still other places where `CMAKE_C_COMPILER` is called which is now `cgcc`, where in fact the real C compiler should be called.
Hence one of the reasons I think #43776 should not have been merged, as I pointed out.
Users will waste a lot of time on chasing subtle bugs.

**Review by lyakh — DISMISSED** [review]
it *looks* good to me, but I certainly cannot verify that this also works for the SOF use-case, I'm relying on @marc-hb for that

**Comment by marc-hb** [comment]
This PR now conflicts with 0b845cd28c47ec276. Temporarily reverting 0b845cd28c47ec276 makes it cherry-pickable and usable again.

**Comment by marc-hb** [comment]
I found yet another bug fixed by cherry-picking this PR (on top of a5994dbd1a5f52 + revert above).

Without this PR, `touch ../zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi` breaks the incremental build. With this PR the incremental build is fine.

Tested-by: Marc Herbert x2

More specifically, without this PR:
```
cd sof/
./scripts/xtensa-build-zephyr.py  --pristine --cmake-args=-DSPARSE=y mtl
ninja -C ../build-mtl/ # OK
touch ../zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi
ninja -C ../build-mtl/       # fails as shown below
ninja -C ../build-mtl/ clean # fails as shown below!!
```

A `--pristine` build is required.

`touch`ing a .c file seems OK, only a `.dtsi` file seems to trigger this bug.

```
-- Found BOARD.dts: zephyr/boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm.dts
command-line: note: in included file (through zephyr/dts/xtensa/xtensa.dtsi, zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi, ...):
zephyr/dts/common/skeleton.dtsi:10:10: error: unrecognized preprocessor line 'address-cells = <1>;'
zephyr/dts/common/skeleton.dtsi:11:10: error: unrecognized preprocessor line 'size-cells = <1>;'
command-line: note: in included file (through zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi, zephyr/boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm.dts, ...):
zephyr/dts/xtensa/xtensa.dtsi:11:18: error: unrecognized preprocessor line 'address-cells = <1>;'
zephyr/dts/xtensa/xtensa.dtsi:12:18: error: unrecognized preprocessor line 'size-cells = <1>;'
command-line: note: in included file (through zephyr/boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm.dts, builtin):
zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:12:18: error: unrecognized preprocessor line 'address-cells = <1>;'
zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:13:18: error: unrecognized preprocessor line 'size-cells = <0>;'
zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:83:18: error: unrecognized preprocessor line 'clock-cells = <0>;'
zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:101:18: error: unrecognized preprocessor line 'clock-cells = <0>;'
zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:107:18: error: unrecognized preprocessor line 'clock-cells = <0>;'
zephyr/dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:122:26: error: unrecognized 
.
.
.
.
```

**Comment by nashif** [comment]
@tejlmand  can you rebase please?

**Comment by tejlmand** [comment]
> @tejlmand can you rebase please?

Done, and documentation for this feature has been added (and updated for sparse).

**Comment by marc-hb** [comment]
Successfully tested in SOF CI here: https://github.com/thesofproject/sof/actions/runs/3897627227/jobs/6655492827
https://github.com/thesofproject/sof/pull/6942

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@stephanosio  anything blocking this? IMO we need this in 3.3 if there are no major blockers.

**Review by stephanosio — APPROVED** [review]

**Review by shubhamtotade — COMMENTED** [review]
Hi @tejlmand 
I want to implement build infrastructure for supporting klocwork tool and add klocwork support for new SCA infrastructure. How to do this?

**Comment by marc-hb** [comment]
@shubhamtotade what have you tried yet?

**Comment by shubhamtotade** [comment]
Hi @marc-hb
I am working on project with zephyr 3.0.99
So though of cherrypicking these 4 commits from #52671 and port it to my project and modify sparse support to klocwork support
1. [cmake: implement build infrastructure for supporting SCA tools.] - As it is
2. [cmake: add sparse support to the new SCA infrastructure]  - 
Made some changes In cmake/sca/klocwork/sca.cmake 
Made some changes in klocwork.template
But while compiling my projectm its giving error 
Build files have been written to: /workdir/cpucp.proc

west build: building application

[1/245] Preparing syscall dependency handling

[2/245] Generating include/generated/version.h

Zephyr version: 3.0.99 (/workdir/zephyr), build: ve.2.51 [8/245) Building C object zephyr/CMakeFiles/offsets.dir/arch/riscv/core/offsets/offsets.c.obj cmake -E env: unknown option -DCPUCP_MODULE_INIT_LEVEL-POST KERNEL'

[9/245] Generating include/generated/offsets.h FAILED: zephyr/include/generated/offsets.h

cd /workdir/cpucp_proc/zephyr && /usr/bin/python3.8 /workdir/zephyr/scripts/gen_offset_header.py -1/workdir/cpucp_proc/zephyr/CMakeFiles/offsets.dir/./arc h/riscv/core/offsets/offsets.c.obj -o /workdir/cpucp_proc/zephyr/include/generated/offsets.h

Traceback (most recent call last):

File "/workdir/zephyr/scripts/gen_offset_header.py", line 80, in <module>

input file = open(args.input, 'rb')

FileNotFoundError: [Errno 2] No such file or directory: '/workdir/cpucp_proc/zephyr/CMakeFiles/offsets.dir/./arch/riscv/core/offsets/offsets.c.obj'

ninja: build stopped: subcommand failed.

FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build/workdir/cpucp.proc

caught signal:

**Pull request merged**